### PR TITLE
enable macos 86/arm release

### PR DESCRIPTION
**Pull request by winitzki**
The build should create macos-x86_64 and macos-arm64 executables and publish them.

### CODE REVIEW

**Comment by winitzki** [comment]
@mmhat  I could not follow one of your comments in the previous PR:

> Also use runner.arch in the conditional executions here:
 ```
   if [ '${{matrix.os.runner}}' == 'windows-latest' ] && [ "${package}" == 'dhall-nix' ]; then 
             if [ '${{matrix.os.runner}}' == 'windows-latest' ] && [ "${package}" == 'dhall-nix' ]; then 
```

I'm not sure how to modify those conditionals to use `runner.arch`. Right now we have only one runner under windows.

**Comment by winitzki** [comment]
@mmhat  I followed your comments in the previous PR but it seems I'm doing something wrong. There is an error with `main.yml`  saying that `runner` is undefined.

**Comment by mmhat** [comment]
@winitzki Oh I guess I didn't read the docs thoroughly... The `runner` context is of course only available when we have a runner :slightly_smiling_face: At line 35 of the failed job you linked we do not have a runner yet, that part is interpreted before a runner picks up a job of the workflow.

**TLDR;** We can only use the runner context inside the steps; I.e. whenever we could write shell code then the runner context is available there.

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -32,7 +32,7 @@ jobs:
               runner: "ubuntu-latest"
             stack-yaml: "stack.ghc-8.10.yaml"
       fail-fast: false
-    name: ${{ matrix.os.runner }} - ${{ matrix.stack-yaml }}
+    name: ${{ matrix.os.runner }} - ${{matrix.os.runner.arch}} - ${{ matrix.stack-yaml }}
```
Comment (mmhat) [inline .github/workflows/main.yml:35]: According to the docs that should be the only key where we cannot use the `runner` context, but that is ok.
```suggestion
    name: ${{ matrix.os.runner }} - ${{ matrix.stack-yaml }}
```

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -48,9 +48,9 @@ jobs:
       - name: "Cache"
         uses: actions/cache@v3
         with:
-          key: ${{ matrix.os.runner }}-${{ hashFiles(matrix.stack-yaml) }}-${{ env.cache_generation }}
+          key: ${{ matrix.os.runner }}-${{matrix.os.runner.arch}}-${{ hashFiles(matrix.stack-yaml) }}-${{ env.cache_generation }}
```
Comment (mmhat) [inline .github/workflows/main.yml:51]:
```suggestion
          key: ${{ matrix.os.runner }}-${{ runner.arch }}-${{ hashFiles(matrix.stack-yaml) }}-${{ env.cache_generation }}
```

File: .github/workflows/main.yml [hunk 3]
```diff
@@ -48,9 +48,9 @@ jobs:
...
           restore-keys: |
-            ${{ matrix.os.runner }}-
+            ${{ matrix.os.runner }}-${{matrix.os.runner.arch}}-
```
Comment (mmhat) [inline .github/workflows/main.yml:53]:
```suggestion
            ${{ matrix.os.runner }}-${{ runner.arch }}-
```

File: .github/workflows/main.yml [hunk 4]
```diff
@@ -131,14 +131,12 @@ jobs:
...
 
-          architecture="$(uname -m)"
-
           function package() (
             local package="${1}"
             local version="$(stack query locals "${package}" version)"
             shift
             local assets=("${@}")
-            local package_file="${package}-${version}-${architecture}-${{runner.os}}.${{ matrix.os.file-extension }}"
+            local package_file="${package}-${version}-${runner.arch}-${{runner.os}}.${{ matrix.os.file-extension }}"
```
Comment (mmhat) [inline .github/workflows/main.yml:139]:
```suggestion
            local package_file="${package}-${version}-${{ runner.os }}-${{ runner.arch }}.${{ matrix.os.file-extension }}"
```

File: .github/workflows/main.yml [hunk 5]
```diff
@@ -159,5 +157,5 @@ jobs:
         if: ${{ matrix.stack-yaml == 'stack.yaml' }}
         uses: actions/upload-artifact@v3
         with:
-          name: 'dhall-${{runner.os}}.${{matrix.os.file-extension}}'
-          path: 'dhall-*${{runner.os}}.${{matrix.os.file-extension}}'
+          name: 'dhall-${{runner.os}}.${{runner.arch}}.${{matrix.os.file-extension}}'
+          path: 'dhall-*${{runner.os}}.${{runner.arch}}.${{matrix.os.file-extension}}'
```
Comment (mmhat) [inline .github/workflows/main.yml:161]:
```suggestion
          name: 'dhall-${{ runner.os }}-${{ runner.arch }}.${{ matrix.os.file-extension }}'
          path: 'dhall-*${{ runner.os }}-${{ runner.arch }}.${{ matrix.os.file-extension }}'
```

File: .mergify.yml [hunk 6]
```diff
@@ -4,9 +4,9 @@ queue_rules:
   - check-success=hydra
   # Mergify explicitly recommends listing each status check separately:
   # https://docs.mergify.io/conditions/#validating-all-status-checks
-  - check-success=macOS-latest - stack.yaml
-  - check-success=ubuntu-latest - stack.yaml
-  - check-success=windows-latest - stack.yaml
+  - check-success=macOS-latest - ARM64 - stack.yaml
+  - check-success=ubuntu-latest - X64 - stack.yaml
+  - check-success=windows-latest - X64 - stack.yaml
```
Comment (mmhat) [inline .mergify.yml:9]:
```suggestion
  - check-success=macOS-latest - stack.yaml
  - check-success=ubuntu-latest - stack.yaml
  - check-success=windows-latest - stack.yaml
```

**Comment by mmhat** [comment]
@winitzki Nice work!

**Comment by winitzki** [comment]
> @winitzki Nice work!

Thank you for your timely help!
